MSC2965: OAuth 2.0 Authorization Server Metadata discovery (#2965)
matrix-org/matrix-spec-proposals: proposals/2965-auth-metadata.md

# MSC2965: OAuth 2.0 Authorization Server Metadata discovery

This proposal is part of the broader [MSC3861: Next-generation auth for Matrix, based on OAuth 2.0/OIDC][MSC3861].

To be able to initiate an OAuth 2.0 login flow to use a Matrix server, the client needs to know the authorization server metadata, as defined in [RFC8414].

## Proposal

This introduces a new Client-Server API endpoint to discover the authorization server metadata used by the homeserver.

### `GET /auth_metadata`

> A request on this endpoint should return a JSON object containing the authorization server metadata as defined in [RFC8414].

Reviewer: For checklist requirements:

```suggestion
A request on this endpoint should return a JSON object containing the authorization server metadata as defined in [RFC8414].
This endpoint does *not* require authentication, and MAY be rate limited per usual. 
```

For example:

```http

> GET /_matrix/client/v1/auth_metadata

Reviewer 2: Could this endpoint also redirect to the `/.well-known/oauth-authorization-server` document on the issuer?

Reviewer: Redirects generally aren't supported on Matrix APIs - it would be a bit weird to specify one for here. If it's the same information though, and considering it's (probably) without authentication, a smart reverse proxy may be able to simply intercept and serve the alternative resource, provided the formats are the same.

The use of `.well-known` is otherwise covered in the alternatives section.

Accept: application/json
```

```http
HTTP/1.1 200 OK
Content-Type: application/json
Cache-Control: public, max-age=3600
```

```json
{
  "authorization_endpoint": "https://account.example.com/oauth2/auth",
  "token_endpoint": "https://account.example.com/oauth2/token",
  "registration_endpoint": "https://account.example.com/oauth2/clients/register",
  "revocation_endpoint": "https://account.example.com/oauth2/revoke",
  "jwks_uri": "https://account.example.com/oauth2/keys",
  "response_types_supported": ["code"],
  "grant_types_supported": ["authorization_code", "refresh_token"],
  "response_modes_supported": ["query", "fragment"],
  "code_challenge_methods_supported": ["S256"],
  "...": "some fields omitted"
}
```

**Note**: The fields required for the main flow outlined by [MSC3861] and its sub-proposals are:

 - `authorization_endpoint`
 - `token_endpoint`
 - `revocation_endpoint`
 - `response_types_supported` including the value `code`
 - `grant_types_supported` including the values `authorization_code` and `refresh_token`
 - `response_modes_supported` including the values `query` and `fragment`
 - `code_challenge_methods_supported` including the value `S256`

> See individual proposals for more details on each field.

Reviewer: A list of the MSCs which apply here would be appreciated, at least to reference what kind of behaviour is attached. They *should not* become dependencies here though, I think.

Author: Have done that in 2aed234c

In this approach, a new endpoint `/_matrix/client/v1/auth_issuer` would return just the issuer URL:

```http
GET /_matrix/client/v1/auth_issuer
Host: example.com
Accept: application/json
```

```http
HTTP/1.1 200 OK
Content-Type: application/json
```

```json
{
  "issuer": "https://account.example.com/"
}
```

The Matrix client would then discover the OpenID Connect Provider configuration by using [OpenID Connect Discovery].

The downside of this approach is that it requires an extra roundtrip to get the metadata.
It also introduces a dependency on an OpenID Connect specification: [MSC3861] proposals tries to build on OAuth 2.0/IETF standards as much as possible.

### Discovery via [RFC8414] well-known endpoint

[RFC8414] OAuth 2.0 Authorization Server Metadata already defines a standard well-known endpoint, under `.well-known/oauth-authorization-server`.
However, the RFC states that an application leveraging this standard should define its own application-specific endpoint, e.g. `/.well-known/matrix-authorization-server`, and *not* use the `.well-known/oauth-authorization-server` endpoint.

> To avoid confusion with the existing `.well-known/matrix/*` documents, this proposal suggests defining a new C-S API endpoint instead.

Reviewer: we could still use `/.well-known/matrix/auth`, though I suppose the underlying question is whether the information is coming from the *server* or *API*: `.well-known` is before any resolution on the Client-Server API URLs, while `/_matrix` is after. It feels like there's equal arguments for either:

* `.well-known` would be nice for consistency with the RFC
* ... but that puts account management in the hands of the DNS holder (which is probably(?) better when using hosted services)
* ... but then the homeserver might be contacting a completely unrelated account server to verify users/actions/etc because it's not aware of the config change
* ... so we'd use a client-server API endpoint instead
* ... but that looks a bit weird when comparing to the RFC

(adding more rationale to the MSC would be appreciated)

Author: We considered this, but I agree with Erik; this is why we didn't go for that solution

### Discovery via WebFinger

OIDC already has a standard way to discover OP from an identifier: WebFinger.
This is already adopted by Mastodon, and might help solve logging in via 3PIDs like emails.

Sample exchange:

```
GET /.well-known/webfinger?
    resource= mxid:@john:example.com &
    rel= http://openid.net/specs/connect/1.0/issuer
Host: example.com
```

```json
{
  "subject": "mxid:@john:matrix.org",
  "links": [
    {
      "rel": "http://openid.net/specs/connect/1.0/issuer",
      "href": "https://account.example.com"
    }
  ]
}
```

The `mxid` scheme is a bit arbitrary here.
The parameters in the URL should be percent-encoded, this was left unencoded for clarity.

The benefits of this approach are that it is standard and decouples the authentication server from the Matrix server:
different authentication servers could be used by different accounts on the server.

The downsides of this approach are:

- the `.well-known/webfinger` resource is dynamic, which can be harder to host/delegate & might conflict with other services leveraging it like Mastodon
- this does not cover discovering the authentication server for user registration

## Security considerations

> None relevant.

Reviewer: If the MSC adds context around "controller" of resources (namely `.well-known` vs off the homeserver's Client-Server API), those details would be relevant here.

Author: I've talked about the .well-known alternative in another section, didn't feel right in the 'security considerations'. Should I still add something here?

Reviewer: It's implied heavily enough where I don't think it's required to update the security considerations section.

# MSC2965: OAuth 2.0 Authorization Server Metadata discovery

This proposal is part of the broader [MSC3861: Next-generation auth for Matrix, based on OAuth 2.0/OIDC][MSC3861].

To be able to initiate an OAuth 2.0 login flow to use a Matrix server, the client needs to know the authorization server metadata, as defined in [RFC8414].

## Proposal

> This introduces a new Client-Server API endpoint to discover the authorization server metadata used by the homeserver.

Reviewer 3: Can we include a rationale for why we don't hardcode the endpoint URLs to be under a `/_matrix/oauth2` scheme? We still need the metdata discovery for other fields, so to an extent there is a question as to "why not", but still.

Author: I've added that in the alternatives section in 8719e6fc

### `GET /auth_metadata`

A request on this endpoint should return a JSON object containing the authorization server metadata as defined in [RFC8414].
This endpoint does _not_ require authentication, and MAY be rate limited per usual.

For example:

```http
GET /_matrix/client/v1/auth_metadata
Host: example.com
Accept: application/json
```

```http
HTTP/1.1 200 OK
Content-Type: application/json
Cache-Control: public, max-age=3600
```

```json
{
  "authorization_endpoint": "https://account.example.com/oauth2/auth",
  "token_endpoint": "https://account.example.com/oauth2/token",
  "registration_endpoint": "https://account.example.com/oauth2/clients/register",
  "revocation_endpoint": "https://account.example.com/oauth2/revoke",
  "jwks_uri": "https://account.example.com/oauth2/keys",
  "response_types_supported": ["code"],
  "grant_types_supported": ["authorization_code", "refresh_token"],
  "response_modes_supported": ["query", "fragment"],
  "code_challenge_methods_supported": ["S256"],
  "...": "some fields omitted"
}
```

> **Note**: The fields required for the main flow outlined by [MSC3861] and its sub-proposals are:

Reviewer 2: Here we define some required fields for the matrix proposals but I think some clarity is needed on whether requirements from RFC8414 also apply.

- `authorization_endpoint` ([MSC2964])
- `token_endpoint` ([MSC2964])
- `revocation_endpoint` ([MSC4254])
- `registration_endpoint` ([MSC2966])
- `response_types_supported` including the value `code` ([MSC2964])
- `grant_types_supported` including the values `authorization_code` and `refresh_token` ([MSC2964])
- `response_modes_supported` including the values `query` and `fragment` ([MSC2964])
- `code_challenge_methods_supported` including the value `S256` ([MSC2964])

See individual proposals for more details on each field.

### Fallback

If the homeserver does not offer next-generation authentication as described in [MSC3861], this endpoint should return a 404 with the `M_UNRECOGNIZED` error code.

> In this case, clients should fall back to using the User-Interactive Authentication flows instead to authenticate the user.

Author: Isn't `/login` *the* main *User-Interactive Authentication flow*?

> In this case, clients should fall back to using the User-Interactive Authentication flows instead to authenticate the user.

Reviewer 4: `/login` doesn't use the UIA mechanism, it's basically a separate system that doesn't support multiple stages or other interactivity 🙃

> In this case, clients should fall back to using the User-Interactive Authentication flows instead to authenticate the user.

Author: Fair enough. My point here is to point at 'the existing auth and account management APIs', which includes:

 - `/login*`
 - `/refresh`
 - `/register`
 - `/logout*`
 - `/account/deactivate`
 - `/account/password`
 - `/account/3pid/add`
 - `/delete_devices`
 - `DELETE /devices/{id}`

Any suggestion how to phrase that?

> ## Rationale for not using a `.well-known` document

Author: Makes sense, moved that in the alternatives section in c3137914

[RFC8414] suggests using an application-specific well-known endpoint instead of the `.well-known/oauth-authorization-server` endpoint.

Considering the rest of the client-server API, there are two potential locations where this could be hosted:

1. On the server name domain, with well-known delegation, e.g. `https://example.com/.well-known/matrix/auth-metadata`
2. On the client-server API endpoint root, e.g. `https://matrix-client.example.com/.well-known/matrix/auth-metadata`

The first option would require making well-known documents mandatory on the server name domain, with a document that may need to be updated more frequently than existing ones.
This isn't practical for some server deployments, and clients may find it challenging to consistently perform this discovery.

> The second option is also impractical, as all other Matrix APIs on this domain are prefixed with `/_matrix`, and it could easily be confused with the set of well-known documents hosted on the server name domain.

Reviewer 5: I don't understand this -- I think the thread (https://github.com/matrix-org/matrix-spec-proposals/pull/2965/files#r1924524746) talks about before/after finding the client API. That rationale seems to be missing here.

Frankly to me, I'd expect this to be under well-known.

Reviewer 6: maybe "impractical" isn't quite the right word here. It would certainly be confusing AF: the point is that `/.well-known/matrix/client` and `/.well-known/matrix/server` are both expected to be found on the *server name* domain.

I've proposed a clarification at https://github.com/matrix-org/matrix-spec-proposals/pull/2965#discussion_r1989772999 which might help.

```http
GET /_matrix/client/v1/auth_issuer
Host: example.com
Accept: application/json
```

```http
HTTP/1.1 200 OK
Content-Type: application/json
```

```json
{
  "issuer": "https://account.example.com/"
}
```

The Matrix client would then discover the OpenID Connect Provider configuration by using [OpenID Connect Discovery].

The downside of this approach is that it requires an extra roundtrip to get the metadata.
It also introduces a dependency on an OpenID Connect specification: [MSC3861] proposals tries to build on OAuth 2.0/IETF standards as much as possible.

### Discovery via [RFC8414] well-known endpoint

[RFC8414] OAuth 2.0 Authorization Server Metadata already defines a standard well-known endpoint, under `.well-known/oauth-authorization-server`.
However, the RFC states that an application leveraging this standard should define its own application-specific endpoint, e.g. `/.well-known/matrix-authorization-server`, and _not_ use the `.well-known/oauth-authorization-server` endpoint.
To avoid confusion with the existing `.well-known/matrix/*` documents, this proposal suggests defining a new C-S API endpoint instead.

> ### Discovery via the well-known client discovery

Reviewer 6: Nit: it's server discovery rather than client discovery, if anything.

maybe:

```suggestion
### Discovery via existing `.well-known` mechanism

> A previous version of this proposal suggested using the well-known client discovery mechanism to discover the authentication server.
> Clients already discover the homeserver when doing a server discovery via the well-known document.
> 
> A new `m.authentication` field is added to this document to support OpenID Connect Provider (OP) discovery.

Reviewer 6: ```suggestion
A previous version of this proposal suggested using the existing [homeserver discovery mechanism](https://spec.matrix.org/v1.13/client-server-api/#server-discovery) to discover the authentication server.

A new `m.authentication` field is added to the `.well-known` document to support OpenID Connect Provider (OP) discovery.
```

```

```json
{
  "issuer": "https://account.example.com/",
  "authorization_endpoint": "https://account.example.com/oauth2/auth",
  "token_endpoint": "https://account.example.com/oauth2/token",
  "registration_endpoint": "https://account.example.com/oauth2/clients/register",
  "revocation_endpoint": "https://account.example.com/oauth2/revoke",
  "jwks_uri": "https://account.example.com/oauth2/keys",
  "response_types_supported": ["code"],
  "grant_types_supported": ["authorization_code", "refresh_token"],
  "response_modes_supported": ["query", "fragment"],
  "code_challenge_methods_supported": ["S256"],
  "...": "some fields omitted"
}
```

**Note**: The fields required for the main flow outlined by [MSC3861] and its sub-proposals are:

- `issuer` (for compliance with [RFC8414])
- `authorization_endpoint` ([MSC2964])
- `token_endpoint` ([MSC2964])
- `revocation_endpoint` ([MSC4254])
- `registration_endpoint` ([MSC2966])
- `response_types_supported` including the value `code` ([MSC2964])
- `grant_types_supported` including the values `authorization_code` and `refresh_token` ([MSC2964])
- `response_modes_supported` including the values `query` and `fragment` ([MSC2964])

> - `code_challenge_methods_supported` including the value `S256` ([MSC2964])

Reviewer 7: It would help with readability if you included a short description (one sentence or so) to describe what each of these fields does, so that we don't need to flip between so many MSCs.  Also, if you could include the titles of the MSCs (and RFC), that would be helpful too.

### Discovery via [RFC8414] well-known endpoint

[RFC8414: OAuth 2.0 Authorization Server Metadata][RFC8414] already defines a standard well-known endpoint, under `.well-known/oauth-authorization-server`.
However, the RFC states that an application leveraging this standard should define its own application-specific endpoint, e.g. `/.well-known/matrix-authorization-server`, and _not_ use the `.well-known/oauth-authorization-server` endpoint.

Considering the rest of the client-server API, there are two potential locations where this could be hosted:

1. On the server name domain, with well-known delegation, e.g. `https://example.com/.well-known/matrix/auth-metadata`
2. On the client-server API endpoint root, e.g. `https://matrix-client.example.com/.well-known/matrix/auth-metadata`

The first option would require making well-known documents mandatory on the server name domain, with a document that may need to be updated more frequently than existing ones.
This isn't practical for some server deployments, and clients may find it challenging to consistently perform this discovery.

> The second option is also impractical, as all other Matrix APIs on this domain are prefixed with `/_matrix`, and it could easily be confused with the set of well-known documents hosted on the server name domain.

Reviewer 6: ```suggestion
The second option would be very confusing, as all other Matrix APIs on the client-server domain are prefixed with `/_matrix`, whereas the existing `.well-known` documents ([`/.well-known/matrix/client`](https://spec.matrix.org/v1.13/client-server-api/#getwell-knownmatrixclient) and [`/.well-known/matrix/server`](https://spec.matrix.org/v1.13/server-server-api/#getwell-knownmatrixserver)) are hosted on the server name domain.
```
